File: packages/metrics/python/collector.py
"""
Python Metrics Collector
Lightweight in-memory metrics collection with periodic flushing
"""

import asyncio
import time
import json
import psutil
import os
from typing import Dict, List, Optional, Any
from datetime import datetime, timedelta
from collections import defaultdict
from dataclasses import dataclass, asdict

from .types import (
    ExecutionMetrics,
    AggregatedMetrics,
    MetricsCollector as IMetricsCollector,
    ExecutionTracker as IExecutionTracker,
    ExecutionStatus,
    ErrorDetail,
    Language,
    AggregationPeriod,
    RpcMetrics,
    ResourceMetrics
)

class ExecutionTracker(IExecutionTracker):
    """Tracks metrics for a single execution"""
    
    def __init__(
        self,
        execution_id: str,
        collector: 'MetricsCollector',
        tool_id: str,
        session_id: str,
        message_id: str,
        language: Language
    ):
        super().__init__(execution_id)
        self.collector = collector
        self.start_time_ns = time.perf_counter_ns()
        self.process = psutil.Process()
        self.start_memory = self.process.memory_info().rss
        
        self.metrics = ExecutionMetrics(
            tool_id=tool_id,
            execution_id=execution_id,
            session_id=session_id,
            message_id=message_id,
            language=language,
            start_time=datetime.utcnow().isoformat(),
            status='success'  # Default
        )
    
    def record_rpc_metrics(self, metrics: RpcMetrics) -> None:
        """Record RPC metrics"""
        if self.metrics.rpc_metrics:
            # Merge with existing
            for key, value in asdict(metrics).items():
                if value is not None:
                    setattr(self.metrics.rpc_metrics, key, value)
        else:
            self.metrics.rpc_metrics = metrics
    
    def record_resource_metrics(self, metrics: ResourceMetrics) -> None:
        """Record resource metrics"""
        if self.metrics.resource_metrics:
            # Merge with existing
            for key, value in asdict(metrics).items():
                if value is not None:
                    setattr(self.metrics.resource_metrics, key, value)
        else:
            self.metrics.resource_metrics = metrics
    
    def record_custom_metric(self, key: str, value: Any) -> None:
        """Record a custom metric"""
        if self.metrics.custom_metrics is None:
            self.metrics.custom_metrics = {}
        self.metrics.custom_metrics[key] = value
    
    def complete(self, status: ExecutionStatus, error: Optional[ErrorDetail] = None) -> None:
        """Complete the execution tracking"""
        end_time_ns = time.perf_counter_ns()
        self.metrics.end_time = datetime.utcnow().isoformat()
        self.metrics.duration = (end_time_ns - self.start_time_ns) / 1_000_000  # Convert to ms
        self.metrics.status = status
        
        if error:
            self.metrics.error = error
        
        # Record final resource metrics if not already set
        if not self.metrics.resource_metrics:
            self.metrics.resource_metrics = ResourceMetrics()
        
        try:
            # Get CPU usage
            self.metrics.resource_metrics.cpu_usage = self.process.cpu_percent()
            
            # Get memory usage
            current_memory = self.process.memory_info().rss
            self.metrics.resource_metrics.memory_used = current_memory
            self.metrics.resource_metrics.memory_delta = current_memory - self.start_memory
        except (psutil.NoSuchProcess, psutil.AccessDenied):
            pass  # Process might have ended
        
        self.collector.record_execution(self.metrics)

class MetricsCollector(IMetricsCollector):
    """Collects and aggregates execution metrics"""
    
    def __init__(
        self,
        max_buffer_size: int = 1000,
        flush_interval: float = 60.0,  # seconds
        auto_flush: bool = True
    ):
        self.executions: List[ExecutionMetrics] = []
        self.aggregated_cache: Dict[str, AggregatedMetrics] = {}
        self.max_buffer_size = max_buffer_size
        self.flush_interval = flush_interval
        self.flush_task: Optional[asyncio.Task] = None
        self._lock = asyncio.Lock()
        
        if auto_flush:
            self.start_auto_flush()
    
    def start_execution(
        self,
        tool_id: str,
        execution_id: str,
        session_id: str,
        message_id: str,
        language: Language
    ) -> ExecutionTracker:
        """Start tracking a new execution"""
        return ExecutionTracker(
            execution_id=execution_id,
            collector=self,
            tool_id=tool_id,
            session_id=session_id,
            message_id=message_id,
            language=language
        )
    
    def record_execution(self, metrics: ExecutionMetrics) -> None:
        """Record completed execution metrics"""
        asyncio.create_task(self._record_execution_async(metrics))
    
    async def _record_execution_async(self, metrics: ExecutionMetrics) -> None:
        """Async version of record_execution"""
        async with self._lock:
            self.executions.append(metrics)
            
            # Check if we need to flush
            if len(self.executions) >= self.max_buffer_size:
                await self.flush()
    
    def get_metrics(self, tool_id: Optional[str] = None) -> List[ExecutionMetrics]:
        """Get execution metrics, optionally filtered by tool"""
        if not tool_id:
            return self.executions.copy()
        
        return [m for m in self.executions if m.tool_id == tool_id]
    
    def get_aggregated_metrics(
        self,
        tool_id: str,
        language: Language,
        period: AggregationPeriod
    ) -> Optional[AggregatedMetrics]:
        """Get aggregated metrics for a tool"""
        cache_key = f"{tool_id}-{language}-{period}"
        
        # Check cache
        cached = self.aggregated_cache.get(cache_key)
        if cached and self._is_cache_valid(cached, period):
            return cached
        
        # Calculate aggregated metrics
        relevant_metrics = [
            m for m in self.executions
            if m.tool_id == tool_id and m.language == language
        ]
        
        if not relevant_metrics:
            return None
        
        aggregated = self._calculate_aggregated_metrics(
            relevant_metrics, tool_id, language, period
        )
        
        # Cache the result
        self.aggregated_cache[cache_key] = aggregated
        return aggregated
    
    def _calculate_aggregated_metrics(
        self,
        metrics: List[ExecutionMetrics],
        tool_id: str,
        language: Language,
        period: AggregationPeriod
    ) -> AggregatedMetrics:
        """Calculate aggregated metrics"""
        now = datetime.utcnow()
        period_start = self._get_period_start(now, period)
        
        # Filter metrics within period
        period_metrics = []
        for m in metrics:
            start_time = datetime.fromisoformat(m.start_time.replace('Z', '+00:00'))
            if start_time >= period_start and start_time <= now:
                period_metrics.append(m)
        
        # Count by status
        status_counts = defaultdict(int)
        durations = []
        error_breakdown = defaultdict(int)
        
        for metric in period_metrics:
            status_counts[metric.status] += 1
            
            if metric.duration is not None:
                durations.append(metric.duration)
            
            if metric.error:
                error_breakdown[metric.error.code] += 1
        
        # Calculate duration statistics
        durations.sort()
        duration_stats = self._calculate_percentiles(durations)
        
        total = len(period_metrics)
        
        return AggregatedMetrics(
            tool_id=tool_id,
            language=language,
            period=period,
            start_time=period_start.isoformat(),
            end_time=now.isoformat(),
            total_executions=total,
            success_count=status_counts['success'],
            error_count=status_counts['error'],
            timeout_count=status_counts['timeout'],
            cancelled_count=status_counts['cancelled'],
            success_rate=status_counts['success'] / total if total > 0 else 0,
            average_duration=duration_stats['average'],
            min_duration=duration_stats['min'],
            max_duration=duration_stats['max'],
            p50_duration=duration_stats['p50'],
            p90_duration=duration_stats['p90'],
            p95_duration=duration_stats['p95'],
            p99_duration=duration_stats['p99'],
            error_breakdown=dict(error_breakdown)
        )
    
    def _calculate_percentiles(self, values: List[float]) -> Dict[str, float]:
        """Calculate percentile statistics"""
        if not values:
            return {
                'average': 0,
                'min': 0,
                'max': 0,
                'p50': 0,
                'p90': 0,
                'p95': 0,
                'p99': 0
            }
        
        return {
            'average': sum(values) / len(values),
            'min': values[0],
            'max': values[-1],
            'p50': self._percentile(values, 0.5),
            'p90': self._percentile(values, 0.9),
            'p95': self._percentile(values, 0.95),
            'p99': self._percentile(values, 0.99)
        }
    
    def _percentile(self, sorted_values: List[float], p: float) -> float:
        """Calculate percentile value"""
        if not sorted_values:
            return 0
        
        index = int(len(sorted_values) * p)
        return sorted_values[min(index, len(sorted_values) - 1)]
    
    def _get_period_start(self, now: datetime, period: AggregationPeriod) -> datetime:
        """Get the start time for a period"""
        if period == 'minute':
            return now.replace(second=0, microsecond=0)
        elif period == 'hour':
            return now.replace(minute=0, second=0, microsecond=0)
        elif period == 'day':
            return now.replace(hour=0, minute=0, second=0, microsecond=0)
        elif period == 'week':
            days_since_monday = now.weekday()
            start = now - timedelta(days=days_since_monday)
            return start.replace(hour=0, minute=0, second=0, microsecond=0)
        
        return now
    
    def _is_cache_valid(self, cached: AggregatedMetrics, period: AggregationPeriod) -> bool:
        """Check if cached metrics are still valid"""
        now = datetime.utcnow()
        cache_age = now - datetime.fromisoformat(cached.end_time.replace('Z', '+00:00'))
        
        # Maximum age for cache validity
        max_age = {
            'minute': timedelta(seconds=10),
            'hour': timedelta(minutes=1),
            'day': timedelta(minutes=5),
            'week': timedelta(minutes=10)
        }
        
        return cache_age < max_age[period]
    
    async def flush(self) -> None:
        """Flush pending metrics"""
        async with self._lock:
            if not self.executions:
                return
            
            metrics_to_flush = self.executions.copy()
            self.executions = []
        
        try:
            await self._persist_metrics(metrics_to_flush)
        except Exception as e:
            # Re-add metrics on failure
            async with self._lock:
                self.executions = metrics_to_flush + self.executions
            raise e
    
    async def _persist_metrics(self, metrics: List[ExecutionMetrics]) -> None:
        """Persist metrics - override for custom persistence"""
        # Default implementation just logs
        print(f"Flushing {len(metrics)} metrics")
    
    def start_auto_flush(self) -> None:
        """Start automatic flushing"""
        async def flush_loop():
            while True:
                await asyncio.sleep(self.flush_interval)
                try:
                    await self.flush()
                except Exception as e:
                    print(f"Auto-flush error: {e}")
        
        self.flush_task = asyncio.create_task(flush_loop())
    
    def stop(self) -> None:
        """Stop the collector"""
        if self.flush_task:
            self.flush_task.cancel()
            self.flush_task = None

# Global metrics collector instance
_global_collector: Optional[MetricsCollector] = None

def get_global_metrics_collector() -> MetricsCollector:
    """Get the global metrics collector"""
    global _global_collector
    if not _global_collector:
        _global_collector = MetricsCollector()
    return _global_collector

def set_global_metrics_collector(collector: MetricsCollector) -> None:
    """Set the global metrics collector"""
    global _global_collector
    if _global_collector:
        _global_collector.stop()
    _global_collector = collector

File: packages/metrics/typescript/src/collector.ts
/**
 * TypeScript Metrics Collector
 * Lightweight in-memory metrics collection with periodic flushing
 */

import { performance } from 'perf_hooks'
import { EventEmitter } from 'events'
import {
  ExecutionMetrics,
  AggregatedMetrics,
  MetricsCollector as IMetricsCollector,
  ExecutionTracker as IExecutionTracker,
  ExecutionStatus,
  ErrorDetail,
  Language,
  AggregationPeriod,
  RpcMetrics,
  ResourceMetrics
} from './types'

export class ExecutionTracker implements IExecutionTracker {
  private metrics: ExecutionMetrics
  private startMark: number

  constructor(
    public readonly executionId: string,
    private collector: MetricsCollector,
    params: {
      toolId: string
      sessionId: string
      messageId: string
      language: Language
    }
  ) {
    this.startMark = performance.now()
    this.metrics = {
      ...params,
      executionId,
      startTime: new Date().toISOString(),
      status: 'success' // Default, will be updated
    }
  }

  recordRpcMetrics(metrics: Partial<RpcMetrics>): void {
    this.metrics.rpcMetrics = {
      ...this.metrics.rpcMetrics,
      ...metrics
    }
  }

  recordResourceMetrics(metrics: Partial<ResourceMetrics>): void {
    this.metrics.resourceMetrics = {
      ...this.metrics.resourceMetrics,
      ...metrics
    }
  }

  recordCustomMetric(key: string, value: any): void {
    if (!this.metrics.customMetrics) {
      this.metrics.customMetrics = {}
    }
    this.metrics.customMetrics[key] = value
  }

  complete(status: ExecutionStatus, error?: ErrorDetail): void {
    const endMark = performance.now()
    this.metrics.endTime = new Date().toISOString()
    this.metrics.duration = endMark - this.startMark
    this.metrics.status = status
    if (error) {
      this.metrics.error = error
    }

    this.collector.recordExecution(this.metrics)
  }
}

export class MetricsCollector extends EventEmitter implements IMetricsCollector {
  private executions: ExecutionMetrics[] = []
  private aggregatedCache = new Map<string, AggregatedMetrics>()
  private flushTimer?: NodeJS.Timer
  private readonly maxBufferSize: number
  private readonly flushInterval: number

  constructor(options: {
    maxBufferSize?: number
    flushInterval?: number // milliseconds
    autoFlush?: boolean
  } = {}) {
    super()
    this.maxBufferSize = options.maxBufferSize || 1000
    this.flushInterval = options.flushInterval || 60000 // 1 minute

    if (options.autoFlush !== false) {
      this.startAutoFlush()
    }
  }

  startExecution(params: {
    toolId: string
    executionId: string
    sessionId: string
    messageId: string
    language: Language
  }): ExecutionTracker {
    return new ExecutionTracker(params.executionId, this, params)
  }

  recordExecution(metrics: ExecutionMetrics): void {
    this.executions.push(metrics)
    this.emit('execution', metrics)

    // Check if we need to flush
    if (this.executions.length >= this.maxBufferSize) {
      this.flush().catch(error => {
        this.emit('error', error)
      })
    }
  }

  getMetrics(toolId?: string): ExecutionMetrics[] {
    if (!toolId) {
      return [...this.executions]
    }
    return this.executions.filter(m => m.toolId === toolId)
  }

  getAggregatedMetrics(
    toolId: string,
    language: Language,
    period: AggregationPeriod
  ): AggregatedMetrics | null {
    const cacheKey = `${toolId}-${language}-${period}`
    
    // Check cache first
    const cached = this.aggregatedCache.get(cacheKey)
    if (cached && this.isCacheValid(cached, period)) {
      return cached
    }

    // Calculate aggregated metrics
    const relevantMetrics = this.executions.filter(
      m => m.toolId === toolId && m.language === language
    )

    if (relevantMetrics.length === 0) {
      return null
    }

    const aggregated = this.calculateAggregatedMetrics(
      relevantMetrics,
      toolId,
      language,
      period
    )

    // Cache the result
    this.aggregatedCache.set(cacheKey, aggregated)
    return aggregated
  }

  private calculateAggregatedMetrics(
    metrics: ExecutionMetrics[],
    toolId: string,
    language: Language,
    period: AggregationPeriod
  ): AggregatedMetrics {
    const now = new Date()
    const periodStart = this.getPeriodStart(now, period)

    // Filter metrics within the period
    const periodMetrics = metrics.filter(m => {
      const startTime = new Date(m.startTime)
      return startTime >= periodStart && startTime <= now
    })

    // Count by status
    const statusCounts = {
      success: 0,
      error: 0,
      timeout: 0,
      cancelled: 0
    }

    const durations: number[] = []
    const errorBreakdown: Record<string, number> = {}

    for (const metric of periodMetrics) {
      statusCounts[metric.status]++
      
      if (metric.duration !== undefined) {
        durations.push(metric.duration)
      }

      if (metric.error) {
        errorBreakdown[metric.error.code] = (errorBreakdown[metric.error.code] || 0) + 1
      }
    }

    // Calculate duration statistics
    durations.sort((a, b) => a - b)
    const durationStats = this.calculatePercentiles(durations)

    return {
      toolId,
      language,
      period,
      startTime: periodStart.toISOString(),
      endTime: now.toISOString(),
      totalExecutions: periodMetrics.length,
      successCount: statusCounts.success,
      errorCount: statusCounts.error,
      timeoutCount: statusCounts.timeout,
      cancelledCount: statusCounts.cancelled,
      successRate: periodMetrics.length > 0 
        ? statusCounts.success / periodMetrics.length 
        : 0,
      averageDuration: durationStats.average,
      minDuration: durationStats.min,
      maxDuration: durationStats.max,
      p50Duration: durationStats.p50,
      p90Duration: durationStats.p90,
      p95Duration: durationStats.p95,
      p99Duration: durationStats.p99,
      errorBreakdown
    }
  }

  private calculatePercentiles(values: number[]): {
    average: number
    min: number
    max: number
    p50: number
    p90: number
    p95: number
    p99: number
  } {
    if (values.length === 0) {
      return {
        average: 0,
        min: 0,
        max: 0,
        p50: 0,
        p90: 0,
        p95: 0,
        p99: 0
      }
    }

    const sum = values.reduce((a, b) => a + b, 0)
    const average = sum / values.length

    return {
      average,
      min: values[0],
      max: values[values.length - 1],
      p50: this.percentile(values, 0.5),
      p90: this.percentile(values, 0.9),
      p95: this.percentile(values, 0.95),
      p99: this.percentile(values, 0.99)
    }
  }

  private percentile(sortedValues: number[], p: number): number {
    if (sortedValues.length === 0) return 0
    const index = Math.ceil(sortedValues.length * p) - 1
    return sortedValues[Math.max(0, Math.min(index, sortedValues.length - 1))]
  }

  private getPeriodStart(now: Date, period: AggregationPeriod): Date {
    const start = new Date(now)
    
    switch (period) {
      case 'minute':
        start.setSeconds(0, 0)
        break
      case 'hour':
        start.setMinutes(0, 0, 0)
        break
      case 'day':
        start.setHours(0, 0, 0, 0)
        break
      case 'week':
        start.setDate(start.getDate() - start.getDay())
        start.setHours(0, 0, 0, 0)
        break
    }
    
    return start
  }

  private isCacheValid(cached: AggregatedMetrics, period: AggregationPeriod): boolean {
    const now = new Date()
    const cacheAge = now.getTime() - new Date(cached.endTime).getTime()
    
    // Cache validity in milliseconds
    const maxAge = {
      minute: 10000,    // 10 seconds
      hour: 60000,      // 1 minute
      day: 300000,      // 5 minutes
      week: 600000      // 10 minutes
    }
    
    return cacheAge < maxAge[period]
  }

  async flush(): Promise<void> {
    if (this.executions.length === 0) {
      return
    }

    const metricsToFlush = [...this.executions]
    this.executions = []

    try {
      await this.persistMetrics(metricsToFlush)
      this.emit('flush', metricsToFlush.length)
    } catch (error) {
      // Re-add metrics on failure
      this.executions.unshift(...metricsToFlush)
      throw error
    }
  }

  private async persistMetrics(metrics: ExecutionMetrics[]): Promise<void> {
    // Override this method to implement persistence
    // Default implementation just logs
    console.log(`Flushing ${metrics.length} metrics`)
  }

  private startAutoFlush(): void {
    this.flushTimer = setInterval(() => {
      this.flush().catch(error => {
        this.emit('error', error)
      })
    }, this.flushInterval)
  }

  stop(): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer)
      this.flushTimer = undefined
    }
  }
}

// Global metrics collector instance
let globalCollector: MetricsCollector | null = null

export function getGlobalMetricsCollector(): MetricsCollector {
  if (!globalCollector) {
    globalCollector = new MetricsCollector()
  }
  return globalCollector
}

export function setGlobalMetricsCollector(collector: MetricsCollector): void {
  if (globalCollector) {
    globalCollector.stop()
  }
  globalCollector = collector
}